File: trees/maximum_depth_of_binary_tree.rb
require 'thread'

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer}
def max_depth(root)
    q = Queue.new
    return 0 if root.nil?

    q << root
    max_depth = 0
    while !q.empty?
        max_depth += 1
        level_len = q.size

        level_len.times do
            node = q.shift()
            
            q << node.left if node.left
            q << node.right if node.right
        end
    end
    return max_depth
end

File: trees/maximum_depth_of_binary_tree.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        q = deque()
        if not root:
            return 0
        q.append(root)
        maxDepth = 0
        while q:
            maxDepth += 1
            level_len = len(q)

            for _ in range(level_len):
                node = q.popleft()

                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return maxDepth
